hazelcast/src/main/java/com/hazelcast/internal/util/ConcurrencyDetection.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.lang.System.nanoTime;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * Responsible for tracking if concurrency is detected.
+ * <p>
+ * Concurrency is currently triggered at 2 levels:
+ * - number of concurrent invocations
+ * - contention on the io system.
+ */
+public abstract class ConcurrencyDetection {
+
+    private final boolean enabled;
+
+    private ConcurrencyDetection(boolean enabled) {
+        this.enabled = enabled;
+    }
+
+    /**
+     * Checks if any concurrency is detected.
+     * <p>
+     * This call is thread-safe;
+     *
+     * @return true if detected, false otherwise.
+     */
+    public abstract boolean isDetected();
+
+    /**
+     * Called when concurrency is detected.
+     * <p>
+     * This call is thread-safe;
+     */
+    public abstract void onDetected();
+
+    /**
+     * Checks if this ConcurrencyDetection is enabled.
+     * <p>
+     * This is useful if you want to fall back to legacy mode.
+     *
+     * @return true if enabled, false otherwise.
+     */
+    public boolean enabled() {
+        return enabled;
+    }
+
+    public static ConcurrencyDetection createEnabled(long durationMs) {
+        return new EnabledConcurrencyDetection(durationMs);
+    }
+
+    public static ConcurrencyDetection createDisabled() {
+        return new DisabledConcurrencyDetection();
+    }
+
+    /**
+     * The always AlwaysConcurrencyDetection indicates that there is always concurrency,
```

Comment:
`AlwaysConcurrencyDetection` -> ?
`even if there is on concurrency` -> `even if there is no concurrency`
`therefor` -> `therefore`

Reply:
Fixed